File: Sources/CompanionCore/HookConfigMerger.swift
import Foundation

/// Merges Worklings' command-hook entries into a tool's JSON hook config. Both
/// tools we target use the same shape — a top-level `"hooks"` object mapping an
/// event name to an array of `{ "hooks": [ { "type": "command", "command" } ] }`
/// — so one merger serves Claude Code's `settings.json` and Codex's `hooks.json`.
///
/// Pure and total on its inputs, so the "never brick an existing config"
/// contract is provable by checks. It never *erases* structure it does not
/// recognize: a config that is present but not valid JSON, a `"hooks"` value
/// that is not an object, or a mapped event whose value is not the expected
/// array all make `connected` throw rather than overwrite. Ownership is matched
/// at the level of an individual command hook, by the adapter's exact file name,
/// so re-connecting is idempotent, disconnecting removes only our hooks (a
/// sibling hook sharing the same entry survives), and a similarly-named user
/// script (`my-worklings-codex-activity-hook-wrapper`) is never mistaken for
/// ours. The adapter names are Worklings-namespaced precisely so that matching
/// on the file name alone — which is what lets a relocated app bundle still
/// clean up its own hooks — cannot accidentally claim an unrelated executable.
public enum HookConfigMerger {
    public struct Mapping: Sendable, Equatable {
        public let event: String
        public let kind: String
        /// Optional matcher restricting which sub-events fire the hook (e.g.
        /// Claude's Notification fires for many types, only some of which mean
        /// "awaiting the user"). Nil fires on every occurrence of the event.
        public let matcher: String?
        public init(event: String, kind: String, matcher: String? = nil) {
            self.event = event
            self.kind = kind
            self.matcher = matcher
        }
    }

    /// How the hook command is written. Both forms *guard* the adapter with a
    /// `[ -x … ]` existence test, so if the app is deleted (the adapter is gone)
    /// the hook degrades to a silent no-op instead of a launch error — the same
    /// convention dotfile tools use for lines they inject into files they don't
    /// own. Both keep the path shell-safe (a space or metacharacter can never
    /// break or be interpreted).
    public enum CommandStyle: Sendable, Equatable {
        /// For a tool that accepts an argv array (Claude Code). We spawn
        /// `/bin/sh -c '<guard>' sh <path> <kind>` and pass the adapter path as a
        /// quoted positional argument (`$1`), so the shell never re-parses it.
        case execForm
        /// For a tool that accepts only a shell string (Codex). The guard is
        /// written inline and the path single-quoted; a missing adapter prints an
        /// empty JSON object (a valid Stop payload) rather than failing.
        case shellForm
    }

    public enum MergeError: Error, Equatable {
        /// The config is present but not a JSON object — refuse to overwrite it.
        case unparseableConfig
        /// `"hooks"`, or a mapped event's value, is present but not the shape we
        /// understand — refuse rather than erase the user's data.
        case unexpectedStructure
    }

    public static func connected(
        configJSON: Data,
        adapterPath: String,
        mappings: [Mapping],
        style: CommandStyle
    ) throws -> Data {
        var root = try object(from: configJSON)
        var hooks = try hooksObject(from: root)

        for mapping in mappings {
            var entries = try existingEntries(in: hooks, event: mapping.event)
            // Strip our own prior hooks (a sibling hook in a shared entry is
            // preserved) so re-connecting is idempotent, then append ours.
            entries = strippingOurHooks(from: entries, adapterPath: adapterPath)
            entries.append(ourEntry(adapterPath: adapterPath, mapping: mapping, style: style))
            hooks[mapping.event] = entries
        }

        root["hooks"] = hooks
        return try data(from: root)
    }

    /// Returns `configJSON` with only our command hooks removed, at the hook
    /// level: a sibling hook sharing an entry with ours is kept, an entry left
    /// with no hooks is dropped, and an emptied `"hooks"` object is removed.
    /// Unfamiliar structure is left untouched (nothing of ours could live in it,
    /// since `connected` would have refused to write it).
    public static func disconnected(
        configJSON: Data,
        adapterPath: String
    ) throws -> Data {
        var root = try object(from: configJSON)
        guard let hooks = root["hooks"] as? [String: Any] else {
            return try data(from: root)
        }

        var kept: [String: Any] = [:]
        for (event, value) in hooks {
            guard let entries = value as? [[String: Any]] else {
                kept[event] = value // preserve anything not shaped like our entries
                continue
            }
            let remaining = strippingOurHooks(from: entries, adapterPath: adapterPath)
            if !remaining.isEmpty {
                kept[event] = remaining
            }
        }

        root["hooks"] = kept.isEmpty ? nil : kept
        return try data(from: root)
    }

    /// Whether any of our command hooks are present.
    public static func isConnected(configJSON: Data, adapterPath: String) -> Bool {
        guard let root = try? object(from: configJSON),
              let hooks = root["hooks"] as? [String: Any] else {
            return false
        }
        for value in hooks.values {
            guard let entries = value as? [[String: Any]] else { continue }
            for entry in entries {
                guard let hookList = entry["hooks"] as? [[String: Any]] else { continue }
                if hookList.contains(where: { hookIsOurs($0, adapterPath: adapterPath) }) {
                    return true
                }
            }
        }
        return false
    }

    // MARK: - Convenience mappings

    /// Claude Code's full lifecycle. `Notification` carries no content we read,
    /// and is matched to only the types that actually mean the user is being
    /// awaited — not auth/elicitation-result notifications.
    public static let claudeCodeMappings: [Mapping] = [
        Mapping(event: "SessionStart", kind: "workStarted"),
        Mapping(event: "Stop", kind: "taskCompleted"),
        Mapping(
            event: "Notification",
            kind: "awaitingInput",
            matcher: "permission_prompt|idle_prompt|elicitation_dialog|agent_needs_input"
        ),
        Mapping(event: "SessionEnd", kind: "workEnded")
    ]

    /// Codex's lifecycle; it has no documented "awaiting input" event.
    public static let codexMappings: [Mapping] = [
        Mapping(event: "SessionStart", kind: "workStarted"),
        Mapping(event: "Stop", kind: "taskCompleted"),
        Mapping(event: "SessionEnd", kind: "workEnded")
    ]

    // MARK: - Building our entry

    /// The guard script for the argv form: run the adapter only if it still
    /// exists and is executable, passing the kind through. The path and kind are
    /// positional arguments (`$1`, `$2`), never spliced into the script text, so
    /// the shell cannot re-parse or word-split them.
    private static let argvGuardScript = "if [ -x \"$1\" ]; then exec \"$1\" \"$2\"; fi"

    private static func ourEntry(adapterPath: String, mapping: Mapping, style: CommandStyle) -> [String: Any] {
        let hook: [String: Any]
        switch style {
        case .execForm:
            // /bin/sh runs the guard; the path and kind are positional args, so a
            // deleted adapter is a silent no-op and the path needs no quoting.
            hook = [
                "type": "command",
                "command": "/bin/sh",
                "args": ["-c", Self.argvGuardScript, "sh", adapterPath, mapping.kind]
            ]
        case .shellForm:
            // Inline guard in a single shell string: run the (single-quoted) path
            // if it exists, else print an empty JSON object so a deleted adapter
            // still returns a valid, content-free success instead of erroring.
            let quoted = singleQuoted(adapterPath)
            hook = [
                "type": "command",
                "command": "if [ -x \(quoted) ]; then \(quoted) \(mapping.kind); else printf '{}'; fi"
            ]
        }
        var entry: [String: Any] = ["hooks": [hook]]
        if let matcher = mapping.matcher {
            entry["matcher"] = matcher
        }
        return entry
    }

    /// POSIX single-quoting: everything inside is literal; an embedded quote is
    /// closed, escaped, and reopened. Neutralizes spaces and metacharacters.
    private static func singleQuoted(_ value: String) -> String {
        "'" + value.replacingOccurrences(of: "'", with: "'\\''") + "'"
    }

    // MARK: - Reading structure (refuse, never erase)

    private static func hooksObject(from root: [String: Any]) throws -> [String: Any] {
        guard let value = root["hooks"] else { return [:] }
        guard let hooks = value as? [String: Any] else { throw MergeError.unexpectedStructure }
        return hooks
    }

    private static func existingEntries(in hooks: [String: Any], event: String) throws -> [[String: Any]] {
        guard let value = hooks[event] else { return [] }
        guard let entries = value as? [[String: Any]] else { throw MergeError.unexpectedStructure }
        return entries
    }

    // MARK: - Ownership (hook-level, exact file name)

    /// Removes our command hooks from each entry, preserving sibling hooks and
    /// dropping an entry only when it is left with none.
    private static func strippingOurHooks(
        from entries: [[String: Any]],
        adapterPath: String
    ) -> [[String: Any]] {
        var result: [[String: Any]] = []
        for var entry in entries {
            guard let hookList = entry["hooks"] as? [[String: Any]] else {
                result.append(entry) // entries we don't understand are left alone
                continue
            }
            let kept = hookList.filter { !hookIsOurs($0, adapterPath: adapterPath) }
            if kept.count == hookList.count {
                result.append(entry) // nothing of ours here
            } else if !kept.isEmpty {
                entry["hooks"] = kept // a sibling hook remains
                result.append(entry)
            }
            // else: the entry held only our hook(s) — drop it
        }
        return result
    }

    /// Every executable path our command hooks point at, across the whole
    /// config. Empty when none are ours. Lets a caller tell "connected and
    /// pointing at a live adapter" from "connected but the adapter path is dead"
    /// (the app was moved or deleted) — ownership (the file name) and liveness
    /// (does that path still exist) are deliberately separate questions.
    public static func ourHookExecutablePaths(configJSON: Data, adapterPath: String) -> [String] {
        guard let root = try? object(from: configJSON),
              let hooks = root["hooks"] as? [String: Any] else {
            return []
        }
        let target = (adapterPath as NSString).lastPathComponent
        var paths: [String] = []
        for value in hooks.values {
            guard let entries = value as? [[String: Any]] else { continue }
            for entry in entries {
                guard let hookList = entry["hooks"] as? [[String: Any]] else { continue }
                for hook in hookList {
                    for candidate in adapterCandidatePaths(in: hook)
                    where (candidate as NSString).lastPathComponent == target {
                        paths.append(candidate)
                    }
                }
            }
        }
        return paths
    }

    private static func hookIsOurs(_ hook: [String: Any], adapterPath: String) -> Bool {
        let target = (adapterPath as NSString).lastPathComponent
        // Ownership is by the adapter's distinctive file name, found anywhere a
        // hook could name it: the `command` itself (an old exec-form config), a
        // single-quoted word inside the command string (the guarded shell form,
        // and the old shell form), or an element of `args` (the guarded argv
        // form). A moved/reinstalled bundle keeps the file name, so it is still
        // recognized; the name is Worklings-namespaced (`worklings-…-activity-hook`),
        // so a differently-named user script — even one reusing a generic stem
        // like `codex-hook` — is never claimed.
        return adapterCandidatePaths(in: hook).contains {
            ($0 as NSString).lastPathComponent == target
        }
    }

    /// Every string in a hook that could be the adapter path: the whole command,
    /// each shell word of the command (recovering a single-quoted path), and each
    /// `args` element. Ownership and liveness both scan these for our file name.
    private static func adapterCandidatePaths(in hook: [String: Any]) -> [String] {
        var candidates: [String] = []
        if let command = hook["command"] as? String {
            candidates.append(command)                        // old exec form: command is the path
            candidates.append(contentsOf: shellWords(command)) // shell forms: path is a quoted word
        }
        if let args = hook["args"] as? [String] {
            candidates.append(contentsOf: args)               // guarded argv form: path is an arg
        }
        return candidates
    }

    /// Splits a shell command string into words, honoring single quotes, double
    /// quotes, and backslash escapes — enough to recover a single-quoted path
    /// (including one whose value contains an apostrophe, written `'\''`).
    private static func shellWords(_ command: String) -> [String] {
        var words: [String] = []
        var current = ""
        var hasWord = false
        var index = command.startIndex
        let end = command.endIndex

        func take() { index = command.index(after: index) }

        while index < end {
            let character = command[index]
            switch character {
            case " ", "\t", "\n":
                if hasWord { words.append(current); current = ""; hasWord = false }
                take()
            case "'":
                hasWord = true
                take()
                while index < end, command[index] != "'" { current.append(command[index]); take() }
                if index < end { take() } // closing quote
            case "\"":
                hasWord = true
                take()
                while index < end, command[index] != "\"" {
                    if command[index] == "\\", command.index(after: index) < end {
                        take(); current.append(command[index]); take(); continue
                    }
                    current.append(command[index]); take()
                }
                if index < end { take() } // closing quote
            case "\\":
                hasWord = true
                take()
                if index < end { current.append(command[index]); take() }
            default:
                hasWord = true
                current.append(character)
                take()
            }
        }
        if hasWord { words.append(current) }
        return words
    }

    // MARK: - JSON

    private static func object(from data: Data) throws -> [String: Any] {
        let isBlank = data.allSatisfy { $0 == 0x20 || $0 == 0x0A || $0 == 0x0D || $0 == 0x09 }
        if data.isEmpty || isBlank {
            return [:]
        }
        guard let parsed = try? JSONSerialization.jsonObject(with: data),
              let object = parsed as? [String: Any] else {
            throw MergeError.unparseableConfig
        }
        return object
    }

    private static func data(from object: [String: Any]) throws -> Data {
        try JSONSerialization.data(
            withJSONObject: object,
            options: [.prettyPrinted, .sortedKeys]
        )
    }
}

File: godot/worklings/core/connect/HookConfigMerger.cs
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Worklings.Core.Connect;

/// One hook we ask a tool to fire: the tool's event name, the activity kind it
/// means to us, and optionally which sub-events count.
public sealed record HookMapping(string Event, string Kind, string? Matcher = null);

/// How the hook command is written. Both forms **guard** the adapter with an
/// existence test, so if the app is deleted the hook degrades to a silent no-op
/// instead of a launch error — the convention dotfile tools use for lines they
/// inject into files they do not own. Both keep the path shell-safe, so a space
/// or a metacharacter can never break or be reinterpreted.
public enum HookCommandStyle
{
    /// For a tool that accepts an argv array (Claude Code). `/bin/sh -c` runs
    /// the guard and the adapter path arrives as a positional argument, so the
    /// shell never re-parses it.
    ExecForm,

    /// For a tool that accepts only a shell string (Codex). The guard is inline
    /// and the path single-quoted; a missing adapter prints an empty JSON object
    /// — a valid Stop payload — rather than failing.
    ShellForm,
}

public enum HookMergeError
{
    /// The config is present but not a JSON object. Refuse to overwrite it.
    UnparseableConfig,

    /// `hooks`, or a mapped event's value, is present but not the shape we
    /// understand. Refuse rather than erase the user's data.
    UnexpectedStructure,
}

public sealed class HookMergeException : System.Exception
{
    public HookMergeError Error { get; }

    public HookMergeException(HookMergeError error)
        : base(error.ToString()) => Error = error;
}

/// Merges Worklings' command hooks into a tool's JSON hook config.
///
/// Both tools we target use the same shape — a top-level `hooks` object mapping
/// an event name to an array of `{ "hooks": [ { "type": "command", … } ] }` — so
/// one merger serves Claude Code's `settings.json` and Codex's `hooks.json`.
///
/// **Pure and total**, so "never brick an existing config" is provable rather
/// than hoped for. It never erases structure it does not recognise: a config
/// that is not valid JSON, a `hooks` value that is not an object, or a mapped
/// event whose value is not the expected array all make `Connected` throw rather
/// than overwrite.
///
/// **Ownership is matched per hook, by the adapter's exact file name.** That
/// makes reconnecting idempotent, lets disconnecting remove only our hooks while
/// a sibling sharing the entry survives, and lets an app that has been moved
/// still recognise and clean up its own wiring. The adapter names are
/// Worklings-namespaced precisely so matching on a file name cannot accidentally
/// claim someone else's script.
///
/// Ported from Sources/CompanionCore/HookConfigMerger.swift. The JSON layer is
/// `JsonNode` where Swift uses `[String: Any]`; the output is pretty-printed
/// with sorted keys, as Swift's is, though the two writers differ in whitespace.
/// That difference is invisible to both tools and to git, which is why the probe
/// compares parsed structure rather than bytes.
public static class HookConfigMerger
{
    /// Claude Code's full lifecycle. `Notification` fires for many things, and
    /// is matched to only the types that actually mean the user is being awaited
    /// — not auth or elicitation-result notifications.
    public static readonly IReadOnlyList<HookMapping> ClaudeCodeMappings = new[]
    {
        new HookMapping("SessionStart", "workStarted"),
        new HookMapping("Stop", "taskCompleted"),
        new HookMapping("Notification", "awaitingInput",
            "permission_prompt|idle_prompt|elicitation_dialog|agent_needs_input"),
        new HookMapping("SessionEnd", "workEnded"),
    };

    /// Codex's lifecycle. It has no documented "awaiting input" event.
    public static readonly IReadOnlyList<HookMapping> CodexMappings = new[]
    {
        new HookMapping("SessionStart", "workStarted"),
        new HookMapping("Stop", "taskCompleted"),
        new HookMapping("SessionEnd", "workEnded"),
    };

    /// Run the adapter only if it still exists and is executable, passing the
    /// kind through. The path and the kind are positional arguments, never
    /// spliced into the script text, so the shell cannot re-parse or word-split
    /// them.
    private const string ArgvGuardScript = "if [ -x \"$1\" ]; then exec \"$1\" \"$2\"; fi";

    public static byte[] Connected(
        byte[] configJson,
        string adapterPath,
        IReadOnlyList<HookMapping> mappings,
        HookCommandStyle style)
    {
        var root = Root(configJson);
        var hooks = HooksObject(root);

        foreach (var mapping in mappings)
        {
            var entries = ExistingEntries(hooks, mapping.Event);
            // Our own prior hooks are stripped first, so reconnecting is
            // idempotent rather than additive. A sibling hook sharing an entry
            // is preserved.
            entries = StrippingOurHooks(entries, adapterPath);
            entries.Add(OurEntry(adapterPath, mapping, style));
            hooks[mapping.Event] = ToArray(entries);
        }

        root["hooks"] = hooks;
        return Serialise(root);
    }

    /// The config with only our command hooks removed, at the hook level: a
    /// sibling sharing an entry is kept, an entry left with no hooks is dropped,
    /// and an emptied `hooks` object is removed entirely. Unfamiliar structure is
    /// left alone — nothing of ours could be living in it, because `Connected`
    /// would have refused to write it.
    public static byte[] Disconnected(byte[] configJson, string adapterPath)
    {
        var root = Root(configJson);
        if (root["hooks"] is not JsonObject hooks)
        {
            return Serialise(root);
        }

        var kept = new JsonObject();
        foreach (var (evt, value) in Pairs(hooks))
        {
            if (value is not JsonArray array || !LooksLikeEntries(array))
            {
                // Preserved rather than understood.
                kept[evt] = value?.DeepClone();
                continue;
            }
            var remaining = StrippingOurHooks(Entries(array), adapterPath);
            if (remaining.Count > 0)
            {
                kept[evt] = ToArray(remaining);
            }
        }

        if (kept.Count == 0)
        {
            root.Remove("hooks");
        }
        else
        {
            root["hooks"] = kept;
        }
        return Serialise(root);
    }

    /// Whether any of our command hooks are present.
    public static bool IsConnected(byte[] configJson, string adapterPath)
    {
        JsonObject root;
        try
        {
            root = Root(configJson);
        }
        catch (HookMergeException)
        {
            return false;
        }
        if (root["hooks"] is not JsonObject hooks) return false;

        foreach (var (_, value) in Pairs(hooks))
        {
            if (value is not JsonArray array) continue;
            foreach (var entry in array)
            {
                if (entry is not JsonObject e || e["hooks"] is not JsonArray list) continue;
                foreach (var hook in list)
                {
                    if (hook is JsonObject h && HookIsOurs(h, adapterPath)) return true;
                }
            }
        }
        return false;
    }

    /// Every executable path our command hooks point at.
    ///
    /// Empty when none are ours. This is what lets a caller tell "connected and
    /// pointing at a live adapter" from "connected, but that path is dead
    /// because the app moved" — ownership (the file name) and liveness (does the
    /// path still exist) are deliberately separate questions.
    public static IReadOnlyList<string> OurHookExecutablePaths(
        byte[] configJson, string adapterPath)
    {
        var paths = new List<string>();
        JsonObject root;
        try
        {
            root = Root(configJson);
        }
        catch (HookMergeException)
        {
            return paths;
        }
        if (root["hooks"] is not JsonObject hooks) return paths;

        string target = FileName(adapterPath);
        foreach (var (_, value) in Pairs(hooks))
        {
            if (value is not JsonArray array) continue;
            foreach (var entry in array)
            {
                if (entry is not JsonObject e || e["hooks"] is not JsonArray list) continue;
                foreach (var hook in list)
                {
                    if (hook is not JsonObject h) continue;
                    foreach (string candidate in AdapterCandidatePaths(h))
                    {
                        if (FileName(candidate) == target) paths.Add(candidate);
                    }
                }
            }
        }
        return paths;
    }

    // MARK: - Building our entry

    private static JsonObject OurEntry(
        string adapterPath, HookMapping mapping, HookCommandStyle style)
    {
        JsonObject hook;
        if (style == HookCommandStyle.ExecForm)
        {
            // /bin/sh runs the guard; the path and the kind are positional
            // arguments, so a deleted adapter is a silent no-op and the path
            // needs no quoting at all.
            hook = new JsonObject
            {
                ["type"] = "command",
                ["command"] = "/bin/sh",
                ["args"] = new JsonArray("-c", ArgvGuardScript, "sh", adapterPath, mapping.Kind),
            };
        }
        else
        {
            // An inline guard in one shell string: run the single-quoted path if
            // it exists, else print an empty JSON object, so a deleted adapter
            // still returns a valid content-free success instead of erroring.
            string quoted = SingleQuoted(adapterPath);
            hook = new JsonObject
            {
                ["type"] = "command",
                ["command"] =
                    $"if [ -x {quoted} ]; then {quoted} {mapping.Kind}; else printf '{{}}'; fi",
            };
        }

        var entry = new JsonObject { ["hooks"] = new JsonArray(hook) };
        if (mapping.Matcher is string matcher)
        {
            entry["matcher"] = matcher;
        }
        return entry;
    }

    /// POSIX single-quoting: everything inside is literal, and an embedded quote
    /// is closed, escaped and reopened. Neutralises spaces and metacharacters.
    private static string SingleQuoted(string value) =>
        "'" + value.Replace("'", "'\\''") + "'";

    // MARK: - Reading structure (refuse, never erase)

    private static JsonObject HooksObject(JsonObject root)
    {
        if (root["hooks"] is null) return new JsonObject();
        if (root["hooks"] is not JsonObject hooks) throw Refuse(HookMergeError.UnexpectedStructure);
        return (JsonObject)hooks.DeepClone();
    }

    private static List<JsonObject> ExistingEntries(JsonObject hooks, string evt)
    {
        if (hooks[evt] is null) return new List<JsonObject>();
        if (hooks[evt] is not JsonArray array || !LooksLikeEntries(array))
        {
            throw Refuse(HookMergeError.UnexpectedStructure);
        }
        return Entries(array);
    }

    /// Swift's cast to `[[String: Any]]` fails unless *every* element is an
    /// object, and that failure is what makes the merger refuse. C#'s type test
    /// is per element, so the same question has to be asked explicitly.
    private static bool LooksLikeEntries(JsonArray array)
    {
        foreach (var element in array)
        {
            if (element is not JsonObject) return false;
        }
        return true;
    }

    private static List<JsonObject> Entries(JsonArray array)
    {
        var entries = new List<JsonObject>();
        foreach (var element in array)
        {
            // Cloned out of the source tree: a JsonNode has one parent, and
            // moving a live node into the tree being built detaches it from the
            // one being read.
            entries.Add((JsonObject)element!.DeepClone());
        }
        return entries;
    }

    // MARK: - Ownership (per hook, exact file name)

    private static List<JsonObject> StrippingOurHooks(
        List<JsonObject> entries, string adapterPath)
    {
        var result = new List<JsonObject>();
        foreach (var entry in entries)
        {
            if (entry["hooks"] is not JsonArray list)
            {
                result.Add(entry); // entries we do not understand are left alone
                continue;
            }

            var kept = new JsonArray();
            int total = 0;
            foreach (var hook in list)
            {
                total++;
                if (hook is JsonObject h && HookIsOurs(h, adapterPath)) continue;
                kept.Add(hook!.DeepClone());
            }

            if (kept.Count == total)
            {
                result.Add(entry); // nothing of ours here
            }
            else if (kept.Count > 0)
            {
                entry["hooks"] = kept; // a sibling hook remains
                result.Add(entry);
            }
            // else: the entry held only ours — drop it
        }
        return result;
    }

    private static bool HookIsOurs(JsonObject hook, string adapterPath)
    {
        // Ownership is the adapter's distinctive file name, found anywhere a
        // hook could name it: the command itself, a quoted word inside a command
        // string, or an element of args. A moved or reinstalled bundle keeps the
        // file name and is still recognised; the name is Worklings-namespaced,
        // so a differently-named user script is never claimed.
        string target = FileName(adapterPath);
        foreach (string candidate in AdapterCandidatePaths(hook))
        {
            if (FileName(candidate) == target) return true;
        }
        return false;
    }

    /// Every string in a hook that could be the adapter path: the whole command,
    /// each shell word of it (recovering a single-quoted path), and each `args`
    /// element.
    private static List<string> AdapterCandidatePaths(JsonObject hook)
    {
        var candidates = new List<string>();
        if (hook["command"] is JsonValue commandValue
            && commandValue.TryGetValue(out string? command) && command is not null)
        {
            candidates.Add(command);                 // old exec form: the command is the path
            candidates.AddRange(ShellWords(command)); // shell forms: a quoted word
        }
        if (hook["args"] is JsonArray args)
        {
            foreach (var arg in args)
            {
                if (arg is JsonValue value && value.TryGetValue(out string? text)
                    && text is not null)
                {
                    candidates.Add(text);            // guarded argv form: an argument
                }
            }
        }
        return candidates;
    }

    /// The last path component, without asking the platform.
    ///
    /// `Path.GetFileName` splits on the *host's* separator, so on Windows it
    /// would also split a POSIX path on backslashes — and these paths come out
    /// of a config file written on some other machine as often as from this one.
    private static string FileName(string path)
    {
        int slash = path.LastIndexOf('/');
        return slash < 0 ? path : path[(slash + 1)..];
    }

    /// Splits a shell command into words, honouring single quotes, double quotes
    /// and backslash escapes — enough to recover a single-quoted path, including
    /// one whose value contains an apostrophe.
    private static List<string> ShellWords(string command)
    {
        var words = new List<string>();
        var current = new System.Text.StringBuilder();
        bool hasWord = false;
        int i = 0;

        while (i < command.Length)
        {
            char c = command[i];
            switch (c)
            {
                case ' ':
                case '\t':
                case '\n':
                    if (hasWord)
                    {
                        words.Add(current.ToString());
                        current.Clear();
                        hasWord = false;
                    }
                    i++;
                    break;
                case '\'':
                    hasWord = true;
                    i++;
                    while (i < command.Length && command[i] != '\'') current.Append(command[i++]);
                    if (i < command.Length) i++; // closing quote
                    break;
                case '"':
                    hasWord = true;
                    i++;
                    while (i < command.Length && command[i] != '"')
                    {
                        if (command[i] == '\\' && i + 1 < command.Length)
                        {
                            i++;
                            current.Append(command[i++]);
                            continue;
                        }
                        current.Append(command[i++]);
                    }
                    if (i < command.Length) i++; // closing quote
                    break;
                case '\\':
                    hasWord = true;
                    i++;
                    if (i < command.Length) current.Append(command[i++]);
                    break;
                default:
                    hasWord = true;
                    current.Append(c);
                    i++;
                    break;
            }
        }
        if (hasWord) words.Add(current.ToString());
        return words;
    }

    // MARK: - JSON

    private static HookMergeException Refuse(HookMergeError error) => new(error);

    /// A blank file is an empty object, not a parse failure — a tool that has
    /// been installed but never configured leaves one behind.
    private static JsonObject Root(byte[] configJson)
    {
        bool blank = true;
        foreach (byte b in configJson)
        {
            if (b is not (0x20 or 0x0A or 0x0D or 0x09)) { blank = false; break; }
        }
        if (configJson.Length == 0 || blank) return new JsonObject();

        try
        {
            if (JsonNode.Parse(configJson) is JsonObject root)
            {
                return (JsonObject)root.DeepClone();
            }
        }
        catch (JsonException)
        {
            // Falls through to the same refusal as a non-object.
        }
        throw Refuse(HookMergeError.UnparseableConfig);
    }

    private static JsonArray ToArray(List<JsonObject> entries)
    {
        var array = new JsonArray();
        foreach (var entry in entries) array.Add(entry.DeepClone());
        return array;
    }

    /// A snapshot of an object's pairs, so a caller can rebuild it while reading.
    private static List<KeyValuePair<string, JsonNode?>> Pairs(JsonObject o)
    {
        var pairs = new List<KeyValuePair<string, JsonNode?>>();
        foreach (var pair in o) pairs.Add(pair);
        return pairs;
    }

    /// Pretty-printed with **sorted keys**, matching Swift's
    /// `[.prettyPrinted, .sortedKeys]`. Sorted so a reconnect produces the same
    /// file rather than a reshuffled one — the diff a user sees in their own
    /// config should be only what actually changed.
    private static byte[] Serialise(JsonObject root)
    {
        var stream = new System.IO.MemoryStream();
        // The relaxed encoder, deliberately. .NET's default escapes `'`, `\"`,
        // `<`, `>` and `&` as \uXXXX for HTML safety — which is correct for a
        // web payload and wrong for a config file a person opens and reads. A
        // shell-form hook is full of apostrophes, and `\u0027` everywhere would
        // make the user's own settings.json unreadable to defend against a
        // browser that is never going to see it.
        using (var writer = new Utf8JsonWriter(
                   stream, new JsonWriterOptions
                   {
                       Indented = true,
                       Encoder = System.Text.Encodings.Web.JavaScriptEncoder
                           .UnsafeRelaxedJsonEscaping,
                   }))
        {
            WriteSorted(writer, root);
        }
        return stream.ToArray();
    }

    private static void WriteSorted(Utf8JsonWriter writer, JsonNode? node)
    {
        switch (node)
        {
            case JsonObject o:
            {
                writer.WriteStartObject();
                var keys = new List<string>();
                foreach (var pair in o) keys.Add(pair.Key);
                // Ordinal, so the order does not depend on the machine's locale.
                keys.Sort(System.StringComparer.Ordinal);
                foreach (string key in keys)
                {
                    writer.WritePropertyName(key);
                    WriteSorted(writer, o[key]);
                }
                writer.WriteEndObject();
                break;
            }
            case JsonArray a:
                writer.WriteStartArray();
                foreach (var element in a) WriteSorted(writer, element);
                writer.WriteEndArray();
                break;
            case null:
                writer.WriteNullValue();
                break;
            default:
                node.WriteTo(writer);
                break;
        }
    }
}
